### PR TITLE
Handle bounds for ScatterOp shape function

### LINKED ISSUE #749: Handle bounds in the ScatterOp shape function

### PR DESCRIPTION
**Pull request by sdasgup3**
fixes #749 

For the ScatterOp, where the 
(C16) inputs[k] and result[k] have the same type for any k  [0, N)

and this is how the result type is inferred. 
(https://github.com/openxla/stablehlo/blob/81d520685c5cfe2620f995dbb09ac4399d000c9a/stablehlo/dialect/TypeInference.cpp#L2513).
 
IMO, we do not need to do anything special to handle the bounds as the bounds should get copied as part of copying the types. I wrote a test to check that input bounds are indeed propagated. Also I note that I do not have to use `inferReturnTypeComponents`  at all, for this case.

### CODE REVIEW

File: stablehlo/tests/infer_stablehlo.mlir [hunk 1]
```diff
@@ -471,6 +471,33 @@ func.func @scatter(%input_tensor: tensor<200x100x300xf32>,
 
 // -----
 
+// CHECK-LABEL: func @scatter_bounds
+func.func @scatter_bounds(%input_tensor: tensor<200x?x?xf32, #stablehlo.type_extensions<bounds = [?, 101, 301]>>,
```
Comment (zhouxin913) [inline stablehlo/tests/infer_stablehlo.mlir:475]: nit: the last 2 dims are testing the same thing, how about we change to:
tensor<200x?x?xf32, #stablehlo.type_extensions<bounds = [?, ?, 301]>>

**Review by zhouxin913 — APPROVED** [review]
Approve with a small nit suggestion.

**Comment by zhouxin913** [comment]
@smit-hinsu I believe this op is trivial too. Just add you here for visibility.

File: stablehlo/tests/infer_stablehlo.mlir (hunk 1 above)
Comment (sdasgup3) [reply stablehlo/tests/infer_stablehlo.mlir]: Good point!

**Pull request merged**